department-of-veterans-affairs/vets-api
# Dv 1237 separate user identities

## Description of change
When a user goes through the oauth login flow, our saml-proxy asks the OpenID Auth module to perform an MVI lookup in order to get the user's ICN. That lookup could populate the redis cache for the `UserIdentity` model with a set of attributes slightly different than the attributes stored after a user went through the SSO identity mixing. That would mean that the user would be at risk of seeing different information about themselves on va.gov and/or an OpenID client app depending on which login flow they went through first.

This PR makes the aforementioned MVI lookup and the OpenID token validation code rely on a new `OpenidUser` model so as to keep the user's identity information separate, allowing them to maintain independent sessions between va.gov versus OpenID client apps.

Ticket here: https://github.com/department-of-veterans-affairs/vets-contrib/issues/1237

## Testing done

None yet because I'm working out a local login issue.

## Testing planned

1. Login locally via va.gov SSO
1. Inspect redis keys, ensure only a User and UserIdentity have been stored.
1. Login to a local OpenID client app
1. Inspect redis keys, ensure that a new OpenidUser and OpenidUserIdentity have been stored.
1. Clear redis.
1. Login to a local OpenID client app
1. Inspect redis keys, ensure that only an OpenidUser and OpenidUserIdentity have been stored.
1. Login locally via va.gov SSO
1. Inspect redis keys, ensure that a new User and UserIdentity have been stored.

A similar test case is detailed in the ticket regarding MHV users with multiple MVI users. I'm still tracking down someone who has a PIV card to test this with an appropriate MHV test user.
